File: src/main/resources/public/_tmp/BattleField.js

let doubleArraySum = (arr) => {
    let sum = 0;
    arr.forEach((a) => {
        a.forEach((e) => sum += e);
    });
    return sum;
}

const VALID_SUM_VALUE = 20;

let battleFieldChecker = {
    field: undefined,
    shipsCheckedPositions: undefined,
    shipsCount: undefined,
    init: function(field) {
        this.field = field;
        this.shipsCheckedPositions = [];
        this.shipsCount = [];
        for (let i = 0; i < 10; i++) {
            this.shipsCheckedPositions.push(new Array(10).fill(0));
        }
        for (let i = 4; i >= 1; i--) this.shipsCount.push(i);
    },
    isInvalidPoint: function(x, y) {
        if (this.field[x][y] === 0) return false;
        return this.diagonalInvalid(x, y) || this.shipInvalid(x, y);
    },
    diagonalInvalid: function(x, y) {
        if (x < 9) {
            if (y < 9) {
                if (this.field[x + 1][y + 1] === 1) return true;
            }
            if (y > 0) {
                if (this.field[x + 1][y - 1] === 1) return true;
            }
        }
        return false;
    },
    shipInvalid: function (x, y) {
        let ver = false;
        let hor = false;

        if (x < 9) hor = this.field[x + 1][y] === 1;
        if (y < 9) ver = this.field[x][y + 1] === 1;

        if (hor && ver) return true;

        if (this.shipsCheckedPositions[x][y] === 0) {
            let len = 0; // should be 1, 0 for index in array
            if (hor) {
                for (let _x = x + 1; _x < 10; _x++) {
                    if (this.field[_x][y] === 1) {
                        this.shipsCheckedPositions[_x][y] = 1;
                        len++;
                    } else break;
                }
            } else {
                for (let _y = y + 1; _y < 10; _y++) {
                    if (this.field[x][_y] === 1) {
                        this.shipsCheckedPositions[x][_y] = 1;
                        len++;
                    } else break;
                }
            }

            if (len > 3) return true;

            this.shipsCount[len]--;
            if (this.shipsCount[len] < 0) return true;
        }

        return false;
    }
}

function validateBattlefield(field) {
    if (doubleArraySum(field) !== VALID_SUM_VALUE) return false;

    battleFieldChecker.init(field);

    for (let x = 0; x < 10; x++) {
        for (let y = 0; y < 10; y++) {
            if (battleFieldChecker.isInvalidPoint(x, y)) return false;
        }
    }

    return battleFieldChecker.isShipsCountValid();
}

console.log(validateBattlefield([
    [1, 0, 0, 0, 0, 1, 1, 0, 0, 0],
    [1, 0, 1, 0, 0, 0, 0, 0, 1, 0],
    [1, 0, 1, 0, 1, 1, 1, 0, 1, 0],
    [1, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 1, 0],
    [0, 0, 0, 0, 1, 1, 1, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 1, 0],
    [0, 0, 0, 1, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 1, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
]));

File: src/main/java/_tmp/BattleField.java
package _tmp;

import java.util.Arrays;

public class BattleField {

    static final int VALID_SUM_VALUE = 20;

    public static boolean fieldValidator(int[][] field) {
        if (Arrays.stream(field).flatMapToInt(Arrays::stream).sum() != VALID_SUM_VALUE) return false;

        BattleField bf = new BattleField(field);

        for (int x = 0; x < 10; x++)
            for (int y = 0; y < 10; y++)
                if (bf.isInvalidPoint(x, y))
                    return false;

        return bf.isShipsCountValid();
    }

    int[][] field;
    int[][] shipsCheckedPositions = new int[10][10];
    int[] shipsCount = {4, 3, 2, 1}; // shipsSize = i + 1;

    private BattleField(int[][] field) {
        this.field = field;
    }

    /* . . . . . . . .
     * . . . d . . . .
     * y . * s . . . .
     * . . s d . . . .
     * . . . . . . . .
     * . . x . . . . .
     *
     * s - checked points for gettings ships
     * d - for invalid checks
     *
     */
    public boolean isInvalidPoint(int x, int y) {
        if (field[x][y] == 0) return false;
        return diagonalInvalid(x, y) || shipInvalid(x, y);
    }

    public boolean diagonalInvalid(int x, int y) {
        if (x < 9) {
            if (y < 9) {
                if (field[x + 1][y + 1] == 1) return true;
            }
            if (y > 0) {
                if (field[x + 1][y - 1] == 1) return true;
            }
        }
        return false;
    }

    public boolean shipInvalid(int x, int y) {
        boolean ver = false;
        boolean hor = false;

        if (x < 9) hor = field[x + 1][y] == 1;
        if (y < 9) ver = field[x][y + 1] == 1;

        if (hor && ver) return true;

        if (shipsCheckedPositions[x][y] == 0) {
            int len = 0; // should be 1, 0 for index in array
            if (hor) {
                for (int _x = x + 1; _x < 10; _x++) {
                    if (field[_x][y] == 1) {
                        shipsCheckedPositions[_x][y] = 1;
                        len++;
                    } else break;
                }
            } else {
                for (int _y = y + 1; _y < 10; _y++) {
                    if (field[x][_y] == 1) {
                        shipsCheckedPositions[x][_y] = 1;
                        len++;
                    } else break;
                }
            }

            if (len > 3) return true;

            shipsCount[len]--;
        }

        return false;
    }

    public boolean isShipsCountValid() {
        for (int c : shipsCount) {
            if (c != 0) return false;
        }
        return true;
    }

}